RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc

```diff
@@ -0,0 +1,317 @@
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/Framework/interface/makeRefToBaseProdFrom.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/BTauReco/interface/JetTag.h"
+
+#include "DataFormats/BTauReco/interface/DeepBoostedJetTagInfo.h"
+
+#include "HeterogeneousCore/SonicTriton/interface/TritonEDProducer.h"
+
+#include "RecoBTag/FeatureTools/interface/deep_helpers.h"
+
+#include <iostream>
+#include <fstream>
+#include <algorithm>
+#include <numeric>
+#include <nlohmann/json.hpp>
+
+using namespace btagbtvdeep;
+
+class ParticleNetSonicJetTagsProducer : public TritonEDProducer<> {
+public:
+  explicit ParticleNetSonicJetTagsProducer(const edm::ParameterSet &);
+  ~ParticleNetSonicJetTagsProducer() override;
+
+  void acquire(edm::Event const &iEvent, edm::EventSetup const &iSetup, Input &iInput) override;
+  void produce(edm::Event &iEvent, edm::EventSetup const &iSetup, Output const &iOutput) override;
+  static void fillDescriptions(edm::ConfigurationDescriptions &);
+
+private:
+  typedef std::vector<reco::DeepBoostedJetTagInfo> TagInfoCollection;
+  typedef reco::JetTagCollection JetTagCollection;
+
+  std::vector<float> center_norm_pad(const std::vector<float> &input,
+                                     float center,
+                                     float scale,
+                                     unsigned min_length,
+                                     unsigned max_length,
+                                     float pad_value = 0,
+                                     float replace_inf_value = 0,
+                                     float min = 0,
+                                     float max = -1);
+  void make_inputs(const reco::DeepBoostedJetTagInfo &taginfo);
+
+  const edm::EDGetTokenT<TagInfoCollection> src_;
+  std::vector<std::string> flav_names_;             // names of the output scores
+  std::vector<std::string> input_names_;            // names of each input group - the ordering is important!
+  std::vector<std::vector<int64_t>> input_shapes_;  // shapes of each input group (-1 for dynamic axis)
+  std::vector<unsigned> input_sizes_;               // total length of each input vector
+  std::unordered_map<std::string, PreprocessParams> prep_info_map_;  // preprocessing info for each input group
+  unsigned batchSize_;
+  bool debug_ = false;
+};
+
+ParticleNetSonicJetTagsProducer::ParticleNetSonicJetTagsProducer(const edm::ParameterSet &iConfig)
+    : TritonEDProducer<>(iConfig, "ParticleNetSonicJetTagsProducer"),
+      src_(consumes<TagInfoCollection>(iConfig.getParameter<edm::InputTag>("src"))),
+      flav_names_(iConfig.getParameter<std::vector<std::string>>("flav_names")),
+      batchSize_(iConfig.getParameter<unsigned>("batchSize")),
+      debug_(iConfig.getUntrackedParameter<bool>("debugMode", false)) {
+  // load preprocessing info
```

Comment:
Hi Kevin, I'm a little unsure of how to implement this.  The two constructors are indeed very similar, but the BoostedJetONNXJetTagsProducer class has an extra data_ array that the ParticleNetSonic one does not.  I could add a data_ array to ParticleNetSonic, but that seems unnecessary if it's just for the sake of a helper function.  Just curious about how you would proceed here

Reply:
I think what @kpedro88 means is that this block of code: https://github.com/fastmachinelearning/cmssw/pull/3/files#diff-7d1f135a5c9eb3d7c2f9a32860a4b2b4e0b699e001aac53921b942e0c938f5d5R63-R119

can be refactored into a new function maybe called `preparePreprocessing` (and located in `deep_helpers`?) and it can be called in both constructors.

Reply:
My idea is something roughly like this:
```cpp
template <typename T>
void constructParticleNetExtra(T* obj) {}
//specialization for BoostedJetONNXJetTagsProducer
template <>
void constructParticleNetExtra<BoostedJetONNXJetTagsProducer>(BoostedJetONNXJetTagsProducer* obj) {
  //do data_ stuff here
}
template <typename T>
void constructParticleNet(T* obj){
  //do constructor stuff here, including calling constructParticleNetExtra()
}
```
And then `constructParticleNet()` should be declared as a [(template) friend function](https://isocpp.org/wiki/faq/templates#template-friends) for both producer classes. (The `constructParticleNetExtra()` trick could be done with an if statement at runtime instead of template specialization at compile time.)

Another possibility would be to make both producers inherit from a common base class that implements the common constructor. This would probably be simpler in some ways, but multiple inheritance can present its own pitfalls.

Reply:
You can see another alternative, where I did this using composition instead of inheritance, at https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/SonicTriton/test/TritonGraphModules.cc

Reply:
As discussed on slack, this is now implemented similar to your composition example (though the helper class is in deep_helpers)